tests/rptest/tests/datalake/datalake_e2e_test.py

```diff
@@ -325,6 +325,68 @@ def assert_location_prefix(rows, prefix: str):
                 f"select path from {table_name}.manifests")
             assert_location_prefix(manifests, table.location())
 
+    @cluster(num_nodes=3)
+    @matrix(cloud_storage_type=supported_storage_types(),
+            catalog_type=supported_catalog_types(),
+            custom_partition_spec=[None, "(timestamp_us)", "(number)"])
+    def test_iceberg_partition_key_file_location(self, cloud_storage_type,
+                                                 catalog_type,
+                                                 custom_partition_spec: str):
+        """
+        Test that the data file location includes the partition key
+        """
+        count = 100
+        with DatalakeServices(self.test_ctx,
+                              redpanda=self.redpanda,
+                              catalog_type=catalog_type,
+                              include_query_engines=[QueryEngineType.SPARK
+                                                     ]) as dl:
+            config = {}
+            if custom_partition_spec:
+                config[
+                    "redpanda.iceberg.partition.spec"] = custom_partition_spec
+
+            dl.create_iceberg_enabled_topic(
+                self.topic_name,
+                partitions=2,
+                iceberg_mode="value_schema_id_prefix",
+                config=config)
+
+            schema = avro.loads(avro_schema_str)
+            producer = AvroProducer(
+                {
+                    'bootstrap.servers': self.redpanda.brokers(),
+                    'schema.registry.url':
+                    self.redpanda.schema_reg().split(",")[0]
+                },
+                default_value_schema=schema)
+            for _ in range(count):
+                t = time.time()
+                record = {"number": int(t), "timestamp_us": int(t * 1000000)}
+                producer.produce(topic=self.topic_name, value=record)
+
+            producer.flush()
+            dl.wait_for_translation(self.topic_name, msg_count=count)
+
+            spark = dl.spark()
+            table_name = f"redpanda.{self.topic_name}"
+
+            def assert_custom_partition_key_path(path):
```

Comment:
I think it would be better to assert on the entire path. I.e. that is is prefixed with `<table location>` + `/data/` + `partition key as path` (split string by `/` then by `%3D` and assert at least the keys for example) + `filename`.

With this PR we are trying to match some external expectations about path formats but we don't have any test yet assertion on the actual external expectation (i.e. that parquet file are at specific locations).

Ideally we would also query the parquet files to ensure that the data is partitioned correctly and we don't write values in the wrong partition files? Or existing tests for partitioning are good enough?

(non-blocker comment)

Reply:
I've changed the test to be more restrictive. I think we need an additional test of the partitioning itself